test/standalone/test_rpc_proxy.py

```diff
@@ -736,70 +824,170 @@ class Service(object):
 
-        config = rabbit_config.copy()
-        config['AMQP_URI'] = toxiproxy.uri
+    @pytest.fixture
+    def reply_listener(self, rpc_proxy):
+        return rpc_proxy.reply_listener
 
-        with ServiceRpcProxy("service", config) as proxy:
+    @pytest.yield_fixture
+    def service_rpc(self, rpc_proxy):
+        with rpc_proxy as proxy:
             yield proxy
 
-    def test_normal(self, service_rpc):
-        assert service_rpc.echo(1) == 1
-        assert service_rpc.echo(2) == 2
+    def test_normal(self, container, service_rpc):
+        assert service_rpc.echo("foo") == "foo"
 
-    def test_down(self, service_rpc, toxiproxy):
-        with toxiproxy.disabled():
+    def test_down(self, reply_listener, service_rpc, toxiproxy):
+        """ Verify we detect and recover from closed sockets.
 
-            # fails to set up initial consumer
-            with pytest.raises(socket.error) as exc_info:
-                service_rpc.echo(1)
-            assert "ECONNREFUSED" in str(exc_info.value)
+        This failure mode closes the socket between the consumer and the
+        rabbit broker.
 
-    def test_timeout(self, service_rpc, toxiproxy):
-        with toxiproxy.timeout(stream="downstream"):
+        Attempting to read from the closed socket raises a socket.error
+        and the connection is re-established.
+        """
+        def reset(args, kwargs, result, exc_info):
+            toxiproxy.enable()
+            return True
 
-            with pytest.raises(IOError) as exc_info:
-                service_rpc.echo(1)
-            assert "Socket closed" in str(exc_info.value)
+        with wait_for_call(
+            reply_listener.consumer, 'on_connection_error', callback=reset
+        ):
+            toxiproxy.disable()
 
-    def test_reuse_when_down(self, service_rpc, toxiproxy):
-        assert service_rpc.echo(1) == 1
+            # connection re-established after one error;
+            # make the request _inside_ the context block because the reply
+            # listener doesn't establish a connection until consuming a result
+            assert service_rpc.echo("foo") == "foo"
 
-        with toxiproxy.disabled():
+    def test_upstream_timeout(self, reply_listener, service_rpc, toxiproxy):
+        """ Verify we detect and recover from sockets timing out.
 
-            with pytest.raises(socket.error) as exc_info:
-                service_rpc.echo(2)
-            assert "ECONNREFUSED" in str(exc_info.value)
+        This failure mode means that the socket between the consumer and the
+        rabbit broker times for out `timeout` milliseconds and then closes.
 
-    def test_reuse_when_recovered(self, service_rpc, toxiproxy):
-        assert service_rpc.echo(1) == 1
+        Attempting to read from the socket after it's closed raises a
+        socket.error and the connection will be re-established. If `timeout`
+        is longer than twice the heartbeat interval, the behaviour is the same
```

Comment:
Not a typo but I think Nameko settled on American English spelling? `behaviour/behavior`?

Reply:
My spelling always seems to end up as "transatlantic" 😀 Will fix.

Reply:
Actually `spelling_lang = 'en_GB'` in the sphinx config. Explains why these weren't flagged by the spell checker already.